### PR TITLE
[ROCm] ROCm platform enhancements and miscellaneous features

### PR DESCRIPTION
**Pull request by fwd4**
This is joint effort by @lfengad @minminsun @fwd4 

ROCm enhancements:
- Enhanced column reduction fusion schedule for ROCm on dptech cases
- Support rocm device info retrieval with new driver API
- ROCm 5.2.0 support with tensorflow-rocm2.9

Misc:
- Add cuda device info retrieval with new driver API
- Improve op type/name blacklist by envvar `TAO_OP_TYPE_CLUSTERING_BLACK_LIST` and `TAO_OP_TYPE_CLUSTERING_BLACK_LIST`

### CODE REVIEW

File: tao_compiler/mlir/xla/ral/ral_helper.h [hunk 1]
```diff
@@ -212,7 +212,7 @@ class TaoRalApiRegistry {
 
  private:
   TaoRalApiRegistry();
-  class Impl;
+  struct Impl;
```
Comment (minminsun) [inline tao_compiler/mlir/xla/ral/ral_helper.h:215]: nit: Is this change necessary?

File: tao_compiler/mlir/disc/transforms/shape_utils.cc [hunk 2]
```diff
@@ -644,6 +644,8 @@ SymbolShape* ShapeAnalysisDeprecated::getShape(Value value) {
   for (int64_t i = 0; i < it->second.rank(); ++i) {
     it->second.setSymbolDim(i, getDim(value, i));
   }
+  auto rank = it->second.rank();
+  assert(rank >= 0 && rank <= 1000);
```
Comment (minminsun) [inline tao_compiler/mlir/disc/transforms/shape_utils.cc:648]: The ShapeAnalysisDeprecated::getShape is still in use?

**Review by minminsun — APPROVED** [review]
LGTM, just some nit questions.

File: tao_compiler/decoupling/BUILD [hunk 3]
```diff
@@ -146,6 +146,8 @@ cc_library(
         "@com_google_absl//absl/strings",
     ] + if_cuda_or_rocm([
         "//tensorflow/compiler/xla/service/gpu:gpu_transfer_manager",
+    ]) + if_rocm_is_configured([
+        "//tensorflow/stream_executor/rocm:rocm_driver",
```
[the remark this answers is not in the document]
Comment (fwd4) [reply tao_compiler/decoupling/BUILD:150]: the compiler now requires device info from rocm driver wrapper
